modules/core/04-channel/keeper/upgrade_test.go

```diff
@@ -939,97 +939,95 @@ func (suite *KeeperTestSuite) TestChanUpgradeOpen() {
-// 				err = path.EndpointB.RecvPacket(packet)
-// 				suite.Require().NoError(err)
-
-// 				err = path.EndpointA.ChanUpgradeInit()
-// 				suite.Require().NoError(err)
-
-// 				err = path.EndpointB.ChanUpgradeTry()
-// 				suite.Require().NoError(err)
-
-// 				err = path.EndpointA.ChanUpgradeAck()
-// 				suite.Require().NoError(err)
-
-// 				// Ack packet to delete packet commitment before calling ChanUpgradeOpen
-// 				err = path.EndpointA.AcknowledgePacket(packet, ibctesting.MockAcknowledgement)
-// 				suite.Require().NoError(err)
-// 			},
-// 			nil,
-// 		},
-// 	}
-
-// 	// Create an initial path used only to invoke ConnOpenInit/ChanOpenInit handlers.
-// 	// This bumps the connection/channel identifiers generated for chain A on the
-// 	// next path used to run the upgrade handshake.
-// 	// See issue 4062.
-// 	path = ibctesting.NewPath(suite.chainA, suite.chainB)
-// 	suite.coordinator.SetupClients(path)
-// 	suite.Require().NoError(path.EndpointA.ConnOpenInit())
-// 	suite.coordinator.SetupConnections(path)
-// 	suite.Require().NoError(path.EndpointA.ChanOpenInit())
-
-// 	for _, tc := range testCases {
-// 		tc := tc
-// 		suite.Run(tc.name, func() {
-// 			suite.SetupTest()
-
-// 			path = ibctesting.NewPath(suite.chainA, suite.chainB)
-// 			suite.coordinator.Setup(path)
-
-// 			path.EndpointB.ChannelConfig.ProposedUpgrade.Fields.Version = mock.UpgradeVersion
-// 			path.EndpointA.ChannelConfig.ProposedUpgrade.Fields.Version = mock.UpgradeVersion
-
-// 			tc.malleate()
-
-// 			proofCounterpartyChannel, _, proofHeight := path.EndpointA.QueryChannelUpgradeProof()
-// 			err := suite.chainA.GetSimApp().IBCKeeper.ChannelKeeper.ChanUpgradeOpen(
-// 				suite.chainA.GetContext(), path.EndpointA.ChannelConfig.PortID, path.EndpointA.ChannelID,
-// 				path.EndpointB.GetChannel().State, proofCounterpartyChannel, proofHeight,
-// 			)
-
-// 			expPass := tc.expError == nil
-// 			if expPass {
-// 				suite.Require().NoError(err)
-// 			} else {
-// 				suite.Require().ErrorIs(err, tc.expError)
-// 			}
-// 		})
-// 	}
-// }
+// TestChanUpgradeOpenCounterPartyStates tests the handshake in all valid counterparty states.
+func (suite *KeeperTestSuite) TestChanUpgradeOpenCounterpartyStates() {
+	var path *ibctesting.Path
+	testCases := []struct {
+		name     string
+		malleate func()
+		expError error
+	}{
+		{
+			"success, counterparty in OPEN",
+			func() {
+				err := path.EndpointA.ChanUpgradeInit()
+				suite.Require().NoError(err)
+
+				err = path.EndpointB.ChanUpgradeTry()
+				suite.Require().NoError(err)
+
+				err = path.EndpointA.ChanUpgradeAck()
+				suite.Require().NoError(err)
+
+				err = path.EndpointB.ChanUpgradeConfirm()
+				suite.Require().NoError(err)
+
+				suite.coordinator.CommitBlock(suite.chainA, suite.chainB)
+				suite.Require().NoError(path.EndpointA.UpdateClient())
+			},
+			nil,
+		},
+		{
+			"success, counterparty in FLUSHCOMPLETE",
```

Comment:
we may be able to retro fit this testcase into the tests already in `TestChanUpgradeOpen`

basically take whats in the malleate here, but reassign the path at the start of malleate using something like:

```go
path = ibctesting.NewPath(suite.chainA, suite.chainB)
suite.coordinator.Setup(path)

// send packet from A, recv on B

// init, try, ack, confirm
```

Reply:
yeah I think you're right! Looks like I refactored the test but it ended up being the same as test cases we already have :D this is a nice suggestions.